nssp/delphi_nssp/run.py

```diff
@@ -0,0 +1,133 @@
+        - "socrata_token": str, authentication for upstream data pull
+    - "archive" (optional): if provided, output will be archived with S3
+        - "aws_credentials": Dict[str, str], AWS login credentials (see S3 documentation)
+        - "bucket_name: str, name of S3 bucket to read/write
+        - "cache_dir": str, directory of locally cached data
+"""
+import time
+from datetime import datetime
+import pdb
+import numpy as np
+import pandas as pd
+from delphi_utils import S3ArchiveDiffer, get_structured_logger, create_export_csv
+from delphi_utils.nancodes import add_default_nancodes
+
+from .constants import GEOS, METRICS, CSV_COLS, SENSORS
+from .pull import pull_nssp_data
+
+import us
+
+
+
+
+def add_needed_columns(df, col_names=None):
+    """Short util to add expected columns not found in the dataset."""
+    if col_names is None:
+        col_names = [
+            "se", 
+            "sample_size",
+            "missing_val",
+            "missing_se",
+            "missing_sample_size"]
+
+    for col_name in col_names:
+        df[col_name] = np.nan
+    df = add_default_nancodes(df)
+    return df
+
+
+def logging(start_time, run_stats, logger):
+    """Boilerplate making logs."""
+    elapsed_time_in_seconds = round(time.time() - start_time, 2)
+    min_max_date = run_stats and min(s[0] for s in run_stats)
+    csv_export_count = sum(s[-1] for s in run_stats)
+    max_lag_in_days = min_max_date and (datetime.now() - min_max_date).days
+    formatted_min_max_date = min_max_date and min_max_date.strftime("%Y-%m-%d")
+    logger.info(
+        "Completed indicator run",
+        elapsed_time_in_seconds=elapsed_time_in_seconds,
+        csv_export_count=csv_export_count,
+        max_lag_in_days=max_lag_in_days,
+        oldest_final_export_date=formatted_min_max_date,
+    )
+
+
+def run_module(params):
+    """
+    Run the indicator.
+
+    Arguments
+    --------
+    params:  Dict[str, Any]
+        Nested dictionary of parameters.
+    """
+    start_time = time.time()
+    logger = get_structured_logger(
+        __name__,
+        filename=params["common"].get("log_filename"),
+        log_exceptions=params["common"].get("log_exceptions", True),
+    )
+    export_dir = params["common"]["export_dir"]
+    socrata_token = params["indicator"]["socrata_token"]
+    # if "archive" in params:
+    #     daily_arch_diff = S3ArchiveDiffer(
+    #         params["archive"]["cache_dir"],
+    #         export_dir,
+    #         params["archive"]["bucket_name"],
+    #         "nchs_mortality",
+    #         params["archive"]["aws_credentials"],
+    #     )
+    #     daily_arch_diff.update_cache()
+
+    run_stats = []
+    ## build the base version of the signal at the most detailed geo level you can get.
+    ## compute stuff here or farm out to another function or file
+    df_pull = pull_nssp_data(socrata_token)
+    sensor_i = 0
+    ## aggregate
+    for metric in METRICS:
```

Comment:
suggestion: for clarity/maintainability, I'd recommend using `enumerate(METRICS)` instead of having a separate `sensor_i` index.